src/scmdata/database.py

```diff
@@ -51,6 +52,248 @@ def _check_is_subdir(root, d):
+        """
+        pass
+
+    @abstractmethod
+    def get(self, filters):
+        """
+        Get all matching keys for a given filter
+
+        Parameters
+        ----------
+        filters: dict of str
+            String filters
+            If a level is missing then all values are fetched
+
+        Returns
+        -------
+        list of str
+            Each item is a key which may contain data which is of interest
+        """
+        pass
+
+
+class NetCDFBackend(DatabaseBackend):
+    """
+    On-disk database handler for outputs from SCMs
+
+    Data is split into groups as specified by :attr:`levels`. This allows for fast
+    reading and writing of new subsets of data when a single output file is no longer
+    performant or data cannot all fit in memory.
+    """
+
+    def get_key(self, sr):
+        """
+        Get key where the data will be stored
+
+        The key is the root directory joined with the other information provided. The filepath
+        is also cleaned to remove spaces and special characters.
+
+        Parameters
+        ----------
+        sr : :class:`scmdata.ScmRun`
+            Data to save
+
+        Raises
+        ------
+        ValueError
+            If non-unique metadata or missing metadata is found for each of :attr:`self.kwargs["levels"]`
+
+        Returns
+        -------
+        str
+            Path in which to save the data without spaces or special characters
+        """
+        levels = {
+            level: sr.get_unique_meta(level, no_duplicates=True).replace(os.sep, "_")
+            for level in self.kwargs["levels"]
+        }
+
+        return self._get_out_filepath(**levels)
+
+    def _get_out_filepath(self, **levels):
+        out_levels = []
+        for level in self.kwargs["levels"]:
+            if level not in levels:
+                raise ValueError("expected value for level: {}".format(level))
+            out_levels.append(str(levels[level]))
+
+        out_path = os.path.join(self.kwargs["root_dir"], *out_levels)
+        out_fname = "__".join(out_levels) + ".nc"
+        out_fname = os.path.join(out_path, out_fname)
+
+        _check_is_subdir(self.kwargs["root_dir"], out_fname)
+
+        return _get_safe_filename(out_fname)
+
+    def save(self, sr):
+        """
+        Save a ScmRun to the database
+
+        The dataset should not contain any duplicate metadata for the
+        database levels
+
+        Parameters
+        ----------
+        sr : :class:`scmdata.ScmRun`
+            Data to save
+
+        Raises
+        ------
+        ValueError
+            If duplicate metadata are present
+
+        Returns
+        -------
+        str
+            Key where the data is saved
+        """
+        key = self.get_key(sr)
+
+        ensure_dir_exists(key)
+        if os.path.exists(key):
+            existing_run = ScmRun.from_nc(key)
+
+            sr = run_append([existing_run, sr])
+
+        # Check for required extra dimensions
```

Comment:
Is it worth altering this so dimensions can be specified by the user?

Reply:
Can also leave it given we can now easily write new custom backends as needed.

Reply:
I'll add kwargs which are passed through

Reply:
Opted to pass  `backend_config`  to the backend when it is initialised. 

Since the reading/writing is coupled, changing the writer will likely change how the files are read so the configuration should be at the backend level rather than a reader/writer

Reply:
nice